### ISSUE TITLE
Low action loss but failed real-world pick on Franka Panda with Pi05 LoRA: possible causes?

### ISSUE THREAD

**Issue by HongyuanChen** [Physical-Intelligence/openpi#906]
I am currently using a Franka Panda robotic arm with a wrist camera and a third-view camera fixed at a constant position. I want to perform a pick-and-place task, but the object to be picked may move within the plane and also have some rotation. Since the end-effector gripper needs to grasp the relatively flat two ends of the object, a certain degree of alignment is required. However, the translation range is actually not very large, likely within about 8 cm × 8 cm, and the rotation is limited to between -60° and 60°.

The robot collects real-time joint angles (q), gripper open/close states, and data from the two cameras through motion commands. All data are then synchronized to the 30 Hz frame rate of the wrist camera. The action is 8-dimensional: the first 7 dimensions correspond to the interpolated joint increment (dq), defined as the next frame minus the current frame, and the 8th dimension represents the absolute gripper open/close state (0/1).

A total of 138 episodes were collected, among which 120 were used as the training set and 18 as the validation set. The test set used the same `norm_stats.json` as the training set. I used Pi05 LoRA with `use_quantile_norm = False`, `batch_size = 16`, and `step = 30,000`. The recorded action loss is as follows:

step | train_loss | val_loss
500 | 0.1108 | 0.1088
10k | 0.0498 | 0.0596
20k | 0.0385 | 0.0370
30k | 0.0312 | 0.0270

However, when I deploy the policy, I first reconstruct the predicted action (dq) into (q), then send it to the Franka robot, and use forward kinematics together with a Cartesian impedance controller to drive the robot. I found that the robot cannot even accomplish the initial pick, and there is a very large discrepancy from the expected behavior. Why might this happen?

To verify that the Cartesian impedance controller itself is not introducing a large tracking error, I also evaluated the error between the commanded motion and the actually executed motion. Specifically, I used the recorded robot end-effector pose, and also estimated the gripper tip pose by assuming a 0.3 m distance from the gripper tip to the end flange. Then I computed the error between the commanded pose and the executed pose for both the robot end-effector and the gripper tip.

For the end-effector position error, the RMSE norm error is **0.0075 m**.
For the gripper tip position error, the RMSE norm error is **0.0085 m**.

These tracking errors are quite small and are on the same order as the positional error observed in the experiment, so it seems unlikely that the Cartesian impedance controller is the main cause of the failure. Therefore, I would like to ask why this might be happening.

https://github.com/user-attachments/assets/0824d552-99c2-43fb-9a97-077f8bbd58ce

<img width="2400" height="1600" alt="Image" src="https://github.com/user-attachments/assets/23a7b846-6771-4aa6-8b84-2ea2de31cbda" />
<img width="2400" height="1600" alt="Image" src="https://github.com/user-attachments/assets/74de9022-5039-4150-bab4-b6b5957b1dd5" />

**Comment by HongyuanChen** [comment]
I would like to add some additional details about the deployment pipeline, in case the issue is related to how the predicted actions are interpreted and executed on the real robot.

The deployment setting uses `action_horizon = 20`. A file of ros-policy-bridge connects the policy environment with ros to transfer data.

However, these predicted actions are not executed directly as raw `dq`. After receiving the current measured joint state (`q_snap`), the bridge accumulates the 20 predicted `dq` values into a sequence of 20 future absolute joint targets. In other words, the actual commands sent to the controller are reconstructed absolute joint positions rather than the original predicted joint increments:
```
Q_new[0] = q_snap + dq[0]
Q_new[1] = Q_new[0] + dq[1]
...
Q_new[i] = Q_new[i-1] + dq[i]
```

In addition, the ros-policy-bridge does not directly replace the old action chunk with the new one when asynchronous inference is used. Instead, if a new chunk arrives before the previous one has been fully executed, the bridge merges the old and new trajectories using a freeze + blend + replace strategy. Around the transition region, it further applies local minimum-jerk smoothing in order to reduce jerk and avoid abrupt switching. Finally, at each 30 Hz control tick, one absolute joint target is sent to the low-level controller.

**Comment by rylinjames** [comment]
low loss but real-robot failure almost always comes down to one of two things: obs normalization mismatch or action chunk execution timing.

for normalization: check that the stats you computed at training time are exactly what's being applied at inference. if you recomputed them or used defaults, even small differences in the proprioception range will cause the policy to see out-of-distribution inputs and output garbage. print raw obs before normalization at inference and compare to your training distribution.

for chunk timing: the async blend strategy (freeze + blend + replace) is sensitive to the ratio between your chunk size and inference latency. if inference is slower on real hardware than in your training/eval setup, you end up executing stale chunks for too long and the behavior diverges.

we hit both of these building reflex-vla (github.com/FastCrest/reflex-vla). happy to dig into which one it is if you can share what your raw action outputs look like at the point of failure.

**Comment by jashshah999** [comment]
Low loss but failed deployment often indicates data quality issues that don't show up in the loss curve — action discontinuities at transitions, biased stats, or inconsistent demos. If your training data is in LeRobot format, `lerobot-doctor score` ranks episodes by quality so you can curate the training set.

```bash
pip install lerobot-doctor
lerobot-doctor score /path/to/data --json
```

https://github.com/jashshah999/lerobot-doctor

**Comment by rylinjames** [comment]
good call on data quality as a root cause. lerobot-doctor looks useful for that layer. the failure mode i was describing sits further down the stack (obs normalization at inference time and chunk timing on real hardware), so worth ruling both out. @HongyuanChen have you been able to print your raw obs at inference yet?
